### ISSUE TITLE
--mode-transparent: support git hash-object command

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad-fuse#51]
Something among `write` operations we should make sure to provide
```
(git-annex)lena:/tmp/mnt/dicoms/dartmouth-phantoms/ANAT-T1W_UNCOMBINED-20200330[master]git-annex
$> git -P --git-dir $PWD/.git hash-object -w --stdin       
123
error: insufficient permission for adding an object to repository database /tmp/mnt/dicoms/dartmouth-phantoms/ANAT-T1W_UNCOMBINED-20200330/.git/objects
fatal: Unable to add (null) to database
```
(there I used `Ctrl-d` after `123`)
whenever it works fine in original dataset location

**Comment by jwodder** [comment]
@yarikoptic I'm not sure what's going on here.  Based on the logs, git calls `mknod` to create a temp file in `.git/objects` with a mode of 0o100444 (which means read-only), and then it tries to open that file with mode 0x100002 (which means read and write), and this results in a permissions error.  The only explanation I can come up with at the moment is that FUSE is somehow stripping bits from the `mknod` call.

**Comment by yarikoptic** [comment]
May be there is some "total pass through" FUSE example you could find and try on, then it could help to rule out fuse library issue and just us either not implementing one of the calls?

**Comment by jwodder** [comment]
@yarikoptic I've gotten this to work with commit ce4adb6 on the gh-48 branch, which implements `create()` and other write ops for just the `.git/` directory when `--mode-transparent` is on.

**Comment by yarikoptic** [comment]
I think I am trying correct version but it seems to not work
```shell
(git-annex)lena:/tmp/mnt/openneuro/ds000001[master]
$> git -P --git-dir $PWD/.git hash-object -w --stdin         
123
error: unable to create temporary file: Read-only file system
fatal: Unable to add (null) to database
```

<details>
<summary>relevant portion from -l DEBUG FWIW</summary> 

```shell
[DEBUG  ] op=getattr for path=/openneuro with args (None,) 
[DEBUG  ] op=getattr for path=/openneuro/ds000001 with args (None,) 
[DEBUG  ] op=getattr for path=/openneuro/ds000001/.git with args (None,) 
[DEBUG  ] op=getattr for path=/openneuro/ds000001/.git/objects with args (None,) 
[DEBUG  ] op=getattr for path=/openneuro/ds000001/.git/objects/info with args (None,) 
[DEBUG  ] getattr(path='/home/yoh/datalad/openneuro/ds000001/.git/objects/info', fh=None) 
[DEBUG  ] File exists; calling os.stat() 
[DEBUG  ] Returning {'st_atime': 1643232847.8294964, 'st_ctime': 1643232847.8294964, 'st_gid': 47522, 'st_mode': 17856, 'st_mtime': 1643232847.8294964, 'st_nlink': 1, 'st_size': 0, 'st_uid': 47521} for /home/yoh/datalad/openneuro/ds000001/.git/objects/info 
[DEBUG  ] op=getattr for path=/openneuro/ds000001/.git/objects/info/alternates with args (None,) 
[DEBUG  ] getattr(path='/home/yoh/datalad/openneuro/ds000001/.git/objects/info/alternates', fh=None) 
[DEBUG  ] File not already open 
[DEBUG  ] /home/yoh/datalad/openneuro/ds000001/.git/objects/info/alternates: path resolved to .git/objects/info/alternates in dataset at /home/yoh/datalad/openneuro/ds000001 
[DEBUG  ] get_file_state: .git/objects/info/alternates 
[DEBUG  ] .git/objects/info/alternates: not under annex 
[DEBUG  ] .git/objects/info/alternates: opening directly 
[DEBUG  ] op=getattr for path=/openneuro/ds000001/.git/objects/pack with args (None,) 
[DEBUG  ] getattr(path='/home/yoh/datalad/openneuro/ds000001/.git/objects/pack', fh=None) 
[DEBUG  ] File exists; calling os.stat() 
[DEBUG  ] Returning {'st_atime': 1643232847.8294964, 'st_ctime': 1643232848.445496, 'st_gid': 47522, 'st_mode': 17856, 'st_mtime': 1643232848.445496, 'st_nlink': 1, 'st_size': 198, 'st_uid': 47521} for /home/yoh/datalad/openneuro/ds000001/.git/objects/pack 
[DEBUG  ] op=getattr for path=/openneuro/ds000001/.git/objects/pack/multi-pack-index with args (None,) 
[DEBUG  ] getattr(path='/home/yoh/datalad/openneuro/ds000001/.git/objects/pack/multi-pack-index', fh=None) 
[DEBUG  ] File not already open 
[DEBUG  ] /home/yoh/datalad/openneuro/ds000001/.git/objects/pack/multi-pack-index: path resolved to .git/objects/pack/multi-pack-index in dataset at /home/yoh/datalad/openneuro/ds000001 
[DEBUG  ] get_file_state: .git/objects/pack/multi-pack-index 
[DEBUG  ] .git/objects/pack/multi-pack-index: not under annex 
[DEBUG  ] .git/objects/pack/multi-pack-index: opening directly 
[DEBUG  ] op=opendir for path=/openneuro/ds000001/.git/objects/pack with args () 
[DEBUG  ] opendir(path='/home/yoh/datalad/openneuro/ds000001/.git/objects/pack') 
[DEBUG  ] Counter = 1001 
[DEBUG  ] op=readdir for path=/openneuro/ds000001/.git/objects/pack with args (1001,) 
[DEBUG  ] readdir(path='/home/yoh/datalad/openneuro/ds000001/.git/objects/pack', fh=1001) 
[DEBUG  ] op=getattr for path=/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.keep with args (None,) 
[DEBUG  ] getattr(path='/home/yoh/datalad/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.keep', fh=None) 
[DEBUG  ] File not already open 
[DEBUG  ] /home/yoh/datalad/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.keep: path resolved to .git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.keep in dataset at /home/yoh/datalad/openneuro/ds000001 
[DEBUG  ] get_file_state: .git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.keep 
[DEBUG  ] .git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.keep: not under annex 
[DEBUG  ] .git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.keep: opening directly 
[DEBUG  ] op=getattr for path=/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.promisor with args (None,) 
[DEBUG  ] getattr(path='/home/yoh/datalad/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.promisor', fh=None) 
[DEBUG  ] File not already open 
[DEBUG  ] /home/yoh/datalad/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.promisor: path resolved to .git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.promisor in dataset at /home/yoh/datalad/openneuro/ds000001 
[DEBUG  ] get_file_state: .git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.promisor 
[DEBUG  ] .git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.promisor: not under annex 
[DEBUG  ] .git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.promisor: opening directly 
[DEBUG  ] op=getattr for path=/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.pack with args (None,) 
[DEBUG  ] getattr(path='/home/yoh/datalad/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.pack', fh=None) 
[DEBUG  ] File exists; calling os.stat() 
[DEBUG  ] Returning {'st_atime': 1643232849.237496, 'st_ctime': 1643232849.237496, 'st_gid': 47522, 'st_mode': 33024, 'st_mtime': 1643232849.237496, 'st_nlink': 1, 'st_size': 324555, 'st_uid': 47521} for /home/yoh/datalad/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.pack 
[DEBUG  ] op=releasedir for path=/openneuro/ds000001/.git/objects/pack with args (1001,) 
[DEBUG  ] op=getattr for path=/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.idx with args (None,) 
[DEBUG  ] getattr(path='/home/yoh/datalad/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.idx', fh=None) 
[DEBUG  ] File exists; calling os.stat() 
[DEBUG  ] Returning {'st_atime': 1643232848.4174962, 'st_ctime': 1643232848.4374962, 'st_gid': 47522, 'st_mode': 33024, 'st_mtime': 1643232848.4174962, 'st_nlink': 1, 'st_size': 72780, 'st_uid': 47521} for /home/yoh/datalad/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.idx 
[DEBUG  ] op=open for path=/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.idx with args (32768,) 
[DEBUG  ] open(path='/home/yoh/datalad/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.idx', flags=0x8000) 
[DEBUG  ] Path exists; opening directly 
[DEBUG  ] op=flush for path=/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.idx with args (7,) 
[DEBUG  ] flush(path='/home/yoh/datalad/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.idx', fh=7) 
[DEBUG  ] Flushing directly 
[DEBUG  ] op=read for path=/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.idx with args (73728, 0, 7) 
[DEBUG  ] read(path='/home/yoh/datalad/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.idx', size=73728, offset=0, fh=7) 
[DEBUG  ] Reading directly 
[DEBUG  ] op=getattr for path=/openneuro/ds000001/.git/objects/19 with args (None,) 
[DEBUG  ] getattr(path='/home/yoh/datalad/openneuro/ds000001/.git/objects/19', fh=None) 
[DEBUG  ] File exists; calling os.stat() 
[DEBUG  ] Returning {'st_atime': 1643737452.2119853, 'st_ctime': 1643737452.2119853, 'st_gid': 47522, 'st_mode': 17856, 'st_mtime': 1643737452.2119853, 'st_nlink': 1, 'st_size': 0, 'st_uid': 47521} for /home/yoh/datalad/openneuro/ds000001/.git/objects/19 
[DEBUG  ] op=getattr for path=/openneuro/ds000001/.git/objects/19/0a18037c64c43e6b11489df4bf0b9eb6d2c9bf with args (None,) 
[DEBUG  ] getattr(path='/home/yoh/datalad/openneuro/ds000001/.git/objects/19/0a18037c64c43e6b11489df4bf0b9eb6d2c9bf', fh=None) 
[DEBUG  ] File not already open 
[DEBUG  ] /home/yoh/datalad/openneuro/ds000001/.git/objects/19/0a18037c64c43e6b11489df4bf0b9eb6d2c9bf: path resolved to .git/objects/19/0a18037c64c43e6b11489df4bf0b9eb6d2c9bf in dataset at /home/yoh/datalad/openneuro/ds000001 
[DEBUG  ] get_file_state: .git/objects/19/0a18037c64c43e6b11489df4bf0b9eb6d2c9bf 
[DEBUG  ] .git/objects/19/0a18037c64c43e6b11489df4bf0b9eb6d2c9bf: not under annex 
[DEBUG  ] .git/objects/19/0a18037c64c43e6b11489df4bf0b9eb6d2c9bf: opening directly 
[DEBUG  ] op=getattr for path=/openneuro/ds000001/.git/objects/19/tmp_obj_Qwmyyw with args (None,) 
[DEBUG  ] getattr(path='/home/yoh/datalad/openneuro/ds000001/.git/objects/19/tmp_obj_Qwmyyw', fh=None) 
[DEBUG  ] File not already open 
[DEBUG  ] /home/yoh/datalad/openneuro/ds000001/.git/objects/19/tmp_obj_Qwmyyw: path resolved to .git/objects/19/tmp_obj_Qwmyyw in dataset at /home/yoh/datalad/openneuro/ds000001 
[DEBUG  ] get_file_state: .git/objects/19/tmp_obj_Qwmyyw 
[DEBUG  ] .git/objects/19/tmp_obj_Qwmyyw: not under annex 
[DEBUG  ] .git/objects/19/tmp_obj_Qwmyyw: opening directly 
[DEBUG  ] op=create for path=/openneuro/ds000001/.git/objects/19/tmp_obj_Qwmyyw with args (33024,) 
[DEBUG  ] op=release for path=/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.idx with args (7,) 
[DEBUG  ] release(path='/home/yoh/datalad/openneuro/ds000001/.git/objects/pack/pack-42dedf37d1a55d77a74b526f0c40e02a6bb48db9.idx', fh=7) 
[DEBUG  ] Closing directly 
[DEBUG  ] op=getattr for path=/openneuro/ds000001/.git/annex with args (None,) 
```
</details>

**Comment by jwodder** [comment]
@yarikoptic Did you include the `--mode-transparent` option to `fusefs`?

**Comment by yarikoptic** [comment]
yes, full invocation: `datalad -l debug fusefs -f -d ~/datalad --mode-transparent /tmp/mnt` and I was in subsubdataset there (`openneuro/ds000001`).

**Comment by jwodder** [comment]
@yarikoptic I think the subdataset aspect is the problem.  The code currently only makes the root `.git/` "transparent".  Should all `.git/` directories in a mounted hierarchy get the same treatment?

**Comment by yarikoptic** [comment]
yes, `.git/` in any subdataset should also be transparent, and I thought they were, but I guess not in all aspects.

**Comment by jwodder** [comment]
@yarikoptic 0387dd3 should fix it.

**Comment by yarikoptic** [comment]
> @yarikoptic [0387dd3](https://github.com/datalad/datalad-fuse/commit/0387dd30c0767816cd45f56a7755a897cdd28a33) should fix it.

I think it now works if that object already exists in `.git/objects` but doesn't work in it is a new one

<details>
<summary>some experimentation</summary> 

```shell
(git-annex)lena:/tmp/mnt/openneuro/ds000001[master]git-annex
$> git -P --git-dir $PWD/.git hash-object -w --stdin
123
190a18037c64c43e6b11489df4bf0b9eb6d2c9bf
(dev3) 1 36457 [1].....................................:Tue 01 Feb 2022 07:12:19 PM EST:.
(git-annex)lena:/tmp/mnt/openneuro/ds000001[master]git-annex
$> git -P --git-dir $PWD/.git hash-object -w --stdin
123

cc12087def83027be7cb8486ab4422fa3dd0741a
(dev3) 1 36457 [1].....................................:Tue 01 Feb 2022 07:12:24 PM EST:.
(git-annex)lena:/tmp/mnt/openneuro/ds000001[master]git-annex
$> git -P --git-dir $PWD/.git hash-object -w --stdin
124
error: unable to create temporary file: Read-only file system
fatal: Unable to add (null) to database
(dev3) 1 36457 ->128 [1].....................................:Tue 01 Feb 2022 07:12:41 PM EST:.
(git-annex)lena:/tmp/mnt/openneuro/ds000001[master]git-annex
$> git -P --git-dir $PWD/.git hash-object -w --stdin
124
error: unable to create temporary file: Read-only file system
fatal: Unable to add (null) to database
(dev3) 1 36457 ->128 [1].....................................:Tue 01 Feb 2022 07:13:50 PM EST:.
(git-annex)lena:/tmp/mnt/openneuro/ds000001[master]git-annex
$> git -P --git-dir $PWD/.git hash-object -w --stdin
124
error: unable to create temporary file: Read-only file system
fatal: Unable to add (null) to database
(dev3) 1 36457 ->128 [1].....................................:Tue 01 Feb 2022 07:14:14 P(d(dev3) 1 36457 ->128 [1].....................................:Tue 01 Feb 2022 07:14:34 PM EST:.
(git-annex)lena:/tmp/mnt/openneuro/ds000001[master]git-annex
$> git -P --git-dir $PWD/.git hash-object --stdin   
124
fc902f4f956b349ac5763a01f37a53ee607614d5

```
</details>

**Comment by yarikoptic** [comment]
ideally the fix should come with a unittest running `git hash-object -w` with the same content twice to see if both operate ok, and that the object becomes known to `git`  after the initial invocation.

**Comment by jwodder** [comment]
@yarikoptic It seems to be working fine for me:

```
$ datalad install ///
$ cd datasets.datalad.org
$ datalad install openneuro
$ cd openneuro
$ datalad install ds000001
$ cd ../..
$ datalad -l debug fusefs -d datasets.datalad.org mount -f --mode-transparent
```

In another window:

```
$ cd mount/openneuro/ds000001
$ git -P --git-dir $PWD/.git hash-object -w --stdin
123
190a18037c64c43e6b11489df4bf0b9eb6d2c9bf
$ git -P --git-dir $PWD/.git hash-object -w --stdin
wer7tugfkuyfkutfk
f3fcf2c1eaf307108eb32db7d59b88cd2c120778
```

**Comment by yarikoptic** [comment]
> @yarikoptic It seems to be working fine for me:

do you mean that it works fine without extra fixes, or now with the further fixes it works fine?

**Comment by jwodder** [comment]
@yarikoptic It works fine without extra fixes.

**Comment by yarikoptic** [comment]
hm, so it means that my setup is somehow special then.  Will try to investigate.

**Comment by yarikoptic** [comment]
will not "investigate" -- no longer can reproduce with current version `0.1.0-30-g99f4c21` of #57 so something fixed it and #57 could indeed close it whenever done.
